src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -0,0 +1,548 @@
+                    tag,
+                    |args: NodeAppArgs<'_>| -> Result<bool, TransitionError> {
+                        let NodeAppArgs { node, mut entry, rel_pos } = args;
+                        let protected = global.borrow().protected_tags.contains_key(&node.tag);
+
+                        // Get old perms, may require initialization from the parent
+                        let perm =
+                            entry.or_insert_with(|| AccessChk::no(node.default_initial_perm));
+                        // Optimization: all transitions are idempotent
+                        // Justified by all_transitions_idempotent and
+                        // foreign_read_is_noop_after_write
+                        // FIXME: update this if the protector policy ever gets changed
+                        if rel_pos.is_foreign() {
+                            if let Some(was_write) = perm.latest_applied_was_write {
+                                if was_write >= T::IS_WRITE {
+                                    return Ok(false);
+                                }
+                            } else {
+                                perm.latest_applied_was_write = Some(T::IS_WRITE);
+                            }
+                        } else {
+                            perm.latest_applied_was_write = None;
+                        }
+                        let old_perm = *perm;
+                        let new_perm = T::after_access(rel_pos, old_perm.data, protected)?;
+                        if old_perm.accessed {
+                            protector_allows_transition(global, node.tag, old_perm.data, new_perm)?
+                        }
+                        perm.data = new_perm;
+                        perm.accessed |= !rel_pos.is_foreign();
+                        Ok(true)
+                    },
+                    |args: ErrHandlerArgs<TransitionError>| -> InterpErrorInfo<'tcx> {
+                        let ErrHandlerArgs {
+                            error: error_kind,
+                            rel_pos: rel_pos_of_access,
+                            faulty_tag,
+                        } = args;
+                        TbError {
+                            faulty_tag,
+                            access_kind: T::NAME,
+                            error_kind,
+                            initial_tag: tag,
+                            rel_pos_of_access,
+                        }
+                        .build()
+                    },
+                )?;
+        }
+        Ok(())
+    }
+}
+
+/// Integration with the BorTag garbage collector
+impl Tree {
+    pub fn remove_unreachable_tags(&mut self, live_tags: &FxHashSet<BorTag>) {
+        assert!(self.keep_only_needed(self.root, live_tags)); // root can't be removed
+    }
+
+    /// Traversal function: returns true iff the tag is necessary. A tag is necessary
+    /// if it is either live or it has necessary children.
```

Comment:
Judging from the name this doesn't just return a bool though it also has some side-effects that might be relevant...